python_modules/libraries/dagster-components/dagster_components/core/component_key.py

```diff
@@ -1,13 +1,21 @@
 import re
-from abc import ABC
+from abc import ABC, abstractmethod
 from dataclasses import dataclass
 from pathlib import Path
 
-COMPONENT_TYPE_REGEX = re.compile(r"^([a-zA-Z0-9_]+)@([a-zA-Z0-9_\.]+)$")
+LOCAL_COMPONENT_IDENTIFIER = "file:"
+
+_COMPONENT_NAME_REGEX = r"[a-zA-Z0-9_]+"
+_LOCAL_NAMESPACE_REGEX = rf"{LOCAL_COMPONENT_IDENTIFIER}[a-zA-Z0-9_\.\/-]+"
+_GLOBAL_NAMESPACE_REGEX = r"[a-zA-Z0-9_\.]+"
+
+COMPONENT_TYPENAME_REGEX = re.compile(
+    rf"^({_COMPONENT_NAME_REGEX})@(({_LOCAL_NAMESPACE_REGEX})|({_GLOBAL_NAMESPACE_REGEX}))$"
+)
 
 
 def _name_and_namespace_from_type(typename: str) -> tuple[str, str]:
-    match = COMPONENT_TYPE_REGEX.match(typename)
+    match = COMPONENT_TYPENAME_REGEX.match(typename)
     if not match:
         raise ValueError(f"Invalid component type name: {typename}")
```

Comment:
We should enhance this by (1) including the valid character set; (2) in the case of local components explaining that the filepath needs to conform to this restricted character set.